frontend/src/pages/NewRun.test.tsx

```diff
@@ -0,0 +1,1160 @@
+    // in PipelineList
+
+    it('shows a page error if fetching the selected pipeline fails', async () => {
+      // Don't actually log to console.
+      // tslint:disable-next-line:no-console
+      console.error = jest.spyOn(console, 'error').mockImplementationOnce(() => null);
+
+      const tree = shallow(<TestNewRun {...generateProps() as any} />);
+      await TestUtils.flushPromises();
+
+      TestUtils.makeErrorResponseOnce(getPipelineSpy, 'test getPipeline error');
+
+      const instance = tree.instance() as TestNewRun;
+      const pipelineId = 'some-pipeline-id';
+      instance._pipelineSelectionChanged(pipelineId);
+      // Confirm pipeline selector
+      await instance._pipelineSelectorClosed(true);
+      await TestUtils.flushPromises();
+
+      expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
+        additionalInfo: 'test getPipeline error',
+        message:
+          `Error: failed to retrieve pipeline with ID: ${pipelineId}.`
+          + ' Click Details for more information.',
+        mode: 'error',
+      }));
+      tree.unmount();
+    });
+  });
+
+  describe('cloning from a run', () => {
+    it('fetches the original run if an ID is present in the query params', async () => {
+      const run = newMockRunDetail().run!;
+      const props = generateProps();
+      props.location.search = `?${QUERY_PARAMS.cloneFromRun}=${run.id}`;
+
+      const tree = shallow(<TestNewRun {...props as any} />);
+      await TestUtils.flushPromises();
+
+      expect(getRunSpy).toHaveBeenCalledTimes(1);
+      expect(getRunSpy).toHaveBeenLastCalledWith(run.id);
+      tree.unmount();
+    });
+
+    it('automatically generates the new run name based on the original run\'s name', async () => {
+      const runDetail = newMockRunDetail();
+      runDetail.run!.name = '-original run-';
+      const props = generateProps();
+      props.location.search = `?${QUERY_PARAMS.cloneFromRun}=${runDetail.run!.id}`;
+
+      getRunSpy.mockImplementation(() => runDetail);
+
+      const tree = shallow(<TestNewRun {...props as any} />);
+      await TestUtils.flushPromises();
+
+      expect(tree.state('runName')).toBe('Clone of -original run-');
+      tree.unmount();
+    });
+
+    it('automatically generates the new clone name if the original run was a clone', async () => {
+      const runDetail = newMockRunDetail();
+      runDetail.run!.name = 'Clone of some run';
+      const props = generateProps();
+      props.location.search = `?${QUERY_PARAMS.cloneFromRun}=${runDetail.run!.id}`;
+
+      getRunSpy.mockImplementation(() => runDetail);
+
+      const tree = shallow(<TestNewRun {...props as any} />);
+      await TestUtils.flushPromises();
+
+      expect(tree.state('runName')).toBe('Clone (2) of some run');
+      tree.unmount();
+    });
+
+    it('uses the query param experiment ID over the one in the original run if an ID is present in both', async () => {
```

Comment:
Can this ever happen?
Perhaps we should sanitize query parameters upon entering NewRun? What if all query parameters are specified (experiment, pipeline id, and run id)? What's the expected behavior?